File: sample-chat-obj-c/sample-chat/Components/ChatScreen/Views/TypingView/TypingView.m
//
//  TypingView.m
//  samplechat
//
//  Created by Injoit on 2/11/20.
//  Copyright © 2020 Quickblox. All rights reserved.
//

#import "TypingView.h"
#import "ChatManager.h"

NSString *const TYPING_ONE = @" is typing...";
NSString *const TYPING_TWO = @" are typing...";
NSString *const TYPING_FOUR = @" and 2 more are typing...";

@interface TypingView ()
@property (strong, nonatomic) ChatManager *chatManager;
@property (strong, nonatomic) UILabel *typingLabel;
@end

@implementation TypingView

- (instancetype)init {
    self = [super init];
    if (self) {
        self.chatManager = [ChatManager instance];
        [self setupViews];
    }
    return self;
}
- (void)setupViews {
    self.backgroundColor = [UIColor colorWithRed:0.96f green:0.96f blue:0.98f alpha:1.0f];
    
    self.typingLabel = [[UILabel alloc] init];
    self.typingLabel.textColor = [UIColor colorWithRed:0.42f green:0.48f blue:0.57f alpha:1.0f];
    self.typingLabel.font = [UIFont italicSystemFontOfSize:13.0f];
    self.typingLabel.textAlignment = NSTextAlignmentLeft;
    [self addSubview:self.typingLabel];
    self.typingLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.typingLabel.leftAnchor constraintEqualToAnchor:self.leftAnchor constant:16.0f].active = YES;
    [self.typingLabel.rightAnchor constraintEqualToAnchor:self.rightAnchor constant:-16.0f].active = YES;
    [self.typingLabel.centerYAnchor constraintEqualToAnchor:self.centerYAnchor].active = YES;
    [self.typingLabel.heightAnchor constraintEqualToConstant:15.0f].active = YES;
}

- (void)setupTypingViewWithOpponentUsersIDs:( NSSet * _Nullable )opponentUsersIDs {
    if (!opponentUsersIDs) {
        return;
    }
    NSString *typingString = @"";
    NSMutableArray *userNames = [NSMutableArray array];
    for (NSNumber *num in opponentUsersIDs) {
        QBUUser *opponentUser = [self.chatManager.storage userWithID:num.unsignedIntegerValue];
        if (opponentUser) {
            NSString *userName = opponentUser.fullName.length ? opponentUser.fullName : opponentUser.login;
            [userNames addObject:userName];
        } else {
            [userNames addObject:@"User"];
        }
    }

    
    switch (opponentUsersIDs.count) {
    case 1:
            typingString = [NSString stringWithFormat:@"%@%@", userNames.firstObject, TYPING_ONE];
            break;
    case 2:
            typingString = [NSString stringWithFormat:@"%@ and %@%@", userNames[0], userNames[1], TYPING_TWO];
            break;
    case 3:
            typingString = [NSString stringWithFormat:@"%@, %@, and %@%@", userNames[0], userNames[1], userNames[2], TYPING_TWO];
            break;
    default:
            typingString = [NSString stringWithFormat:@"%@, %@%@", userNames[0], userNames[1], TYPING_FOUR];
            break;
    }
    self.typingLabel.text = typingString;
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/Classes/Views/TypingView.swift
//
//  TypingView.swift
//  sample-conference-videochat-swift
//
//  Created by Injoit on 11/26/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

import UIKit

struct TypingViewConstant {
    static let typingOne = " is typing..."
    static let typingTwo = " are typing..."
    static let typingFour = " and 2 more are typing..."
}

class TypingView: UIView {
    
    private let chatManager = ChatManager.instance
    
    lazy var typingLabel: UILabel = {
        let typingLabel = UILabel()
        typingLabel.textColor = #colorLiteral(red: 0.4235294118, green: 0.4784313725, blue: 0.5725490196, alpha: 1)
        typingLabel.font = .italicSystemFont(ofSize: 13.0)
        typingLabel.textAlignment = .left
        addSubview(typingLabel)
        typingLabel.translatesAutoresizingMaskIntoConstraints = false
        typingLabel.leftAnchor.constraint(equalTo: self.leftAnchor, constant: 16.0).isActive = true
        typingLabel.rightAnchor.constraint(equalTo: self.rightAnchor, constant: -16.0).isActive = true
        typingLabel.centerYAnchor.constraint(equalTo: self.centerYAnchor).isActive = true
        typingLabel.heightAnchor.constraint(equalToConstant: 15.0).isActive = true
        return typingLabel
    }()
    
    func setupTypingView(_ opponentUsersIDs: Set<UInt>?) {
        guard let opponentsIDs = opponentUsersIDs else {
            return
        }
        var typingString = ""
        var userNames = [String]()
        opponentsIDs.forEach { (userID) in
            guard let opponentUser = self.chatManager.storage.user(withID: userID) else {
                return
            }
            if let userName = opponentUser.fullName {
                userNames.append(userName)
            } else {
                userNames.append("User")
            }
        }
        switch opponentsIDs.count {
        case 1:
            typingString = userNames[0] + TypingViewConstant.typingOne
        case 2:
            typingString = userNames[0] + " and " + userNames[1] + TypingViewConstant.typingTwo
        case 3:
            typingString = userNames[0] + ", " + userNames[1] + " and " + userNames[2] + TypingViewConstant.typingTwo
        default:
            typingString = userNames[0] + ", " + userNames[1] + TypingViewConstant.typingFour
        }
        typingLabel.text = typingString
    }
}
